File: core/src/net/namekdev/memcop/domain/MemorySource.java
package net.namekdev.memcop.domain;

import com.badlogic.gdx.utils.Array;

public class MemorySource {
    public final int validSectorsCount;
    public final int sectorsPerRow;
    public final Array<Sector> sectors;

    public boolean canReadFrom;
    public boolean canWriteTo;
    public boolean canWriteToSpecificIndex;
    public boolean canReadFromSpecificIndex;

    private int lastPos = -1;

    public MemorySource(int sectorsPerRow, Array<Sector> sectors) {
        this.sectorsPerRow = sectorsPerRow;
        this.sectors = sectors;

        int i = 0;
        for (Sector sector : sectors) {
            if (sector.isWritable())
                i++;
        }
        this.validSectorsCount = i;
    }

    public MemorySource(int sectorsPerRow, int totalSize) {
        Array<Sector> sectors = new Array<Sector>(totalSize);
        for (int i = 0; i < totalSize; ++i)
            sectors.add(new Sector());

        this.sectorsPerRow = sectorsPerRow;
        this.sectors = sectors;
        this.validSectorsCount = totalSize;
    }

    public int readValue(int requestedIndex) {
        if (!canReadFrom)
            throw new Error("this memory can't be read from");

        int index = canReadFromSpecificIndex ? requestedIndex : ++lastPos;

        return (byte) sectors.get(index).value();
    }

    public boolean saveByte(int destinationIndex, int value) {
        if (!canWriteTo)
            throw new Error("can't write to this memory");

        int index = canWriteToSpecificIndex ? destinationIndex : ++lastPos;
        Sector sector = sectors.get(index);

        if (sector == null)
            return false;

        sector.write(value);
        return true;
    }

    public int getTotalHeight() {
        int t = sectors.size % sectorsPerRow == 0 ? 0 : 1;
        return sectors.size / sectorsPerRow + t;
    }


    public void reset() {
        lastPos = -1;
        for (Sector sector : sectors)
            sector.reset();
    }
}

File: core/src/net/namekdev/memcop/domain/MemorySource.kt
package net.namekdev.memcop.domain

import net.namekdev.memcop.domain.pojo.MemorySourceInfo

class MemorySource(
    val title: String,
    val sectorsPerRow: Int,
    val sectors: Array<Sector>
) {
    val validSectorsCount: Int

    var canReadFrom: Boolean = false
    var canWriteTo: Boolean = false
    var canWriteToSpecificIndex: Boolean = false
    var canReadFromSpecificIndex: Boolean = false

    var lastPos = -1
        private set

    val totalHeight: Int
        get() {
            val t = if (sectors.size % sectorsPerRow == 0) 0 else 1
            return sectors.size / sectorsPerRow + t
        }

    init {
        var i = 0
        for (sector in sectors) {
            if (sector.isWritable)
                i++
        }
        this.validSectorsCount = i
    }

    fun readValue(requestedIndex: Int): Int {
        if (!canReadFrom)
            throw Error("this memory can't be read from")

        val index = if (canReadFromSpecificIndex) requestedIndex else ++lastPos
        lastPos = index

        return sectors.get(index).value().toByte().toInt()
    }

    fun saveByte(destinationIndex: Int, value: Int): Boolean {
        if (!canWriteTo)
            throw Error("can't write to this memory")

        val index = if (canWriteToSpecificIndex) destinationIndex else ++lastPos
        val sector = sectors.get(index) ?: return false

        lastPos = index
        sector.write(value)
        return true
    }


    fun reset() {
        lastPos = -1
        for (sector in sectors)
            sector.reset()
    }

    companion object {
        fun fromPojo(info: MemorySourceInfo): MemorySource {
            val sectors: Array<Sector> = (
                if (info.sectors != null)
                    info.sectors.map { it -> Sector.fromPojo(it) }.toTypedArray()
                else {
                    if (info.brokenSectorIndices == null) {
                        Array(info.size) { Sector(false) }
                    }
                    else Array(info.size) { i ->
                        val isBroken = info.brokenSectorIndices.contains(i)
                        Sector(isBroken)
                    }
                }
            )

            val mem = MemorySource(info.title, info.sectorsPerRow, sectors)
            mem.canReadFrom = info.canReadFrom
            mem.canReadFromSpecificIndex = info.canReadFromSpecificIndex
            mem.canWriteTo = info.canWriteTo
            mem.canWriteToSpecificIndex = info.canWriteToSpecificIndex

            return mem
        }
    }
}
